### PR TITLE
Added service and node filtering abilities into QueryOptions

### PR DESCRIPTION
**Pull request by declard**
Added [server-side filtering support](https://www.consul.io/api-docs/features/filtering) as an eDSL

This implementation supports filtering by the following components:
- Service
  - ID
  - Tags
  - Meta

With the following operations:
- and
- or
- not
- ==
- !=
- contains
- is empty

For Health.Service

whenever they are applicable to the listed components

Functionally satisfies #26

### CODE REVIEW

**Comment by declard** [comment]
TODO:
- format files
- fix filenames in the comments
- write proper tests

**Comment by arxange1** [comment]
Nice to have, @sergeyshaykhullin could you check this one, please?

**Comment by declard** [comment]
@marcin-krystianc, @jgiannuzzi please advise whom I should ask to take a look at the code

**Comment by jgiannuzzi** [comment]
Thanks a lot for your PR @declard!

We will both take a look at it 😄

**Comment by sergeyshaykhullin** [comment]
Oh, sorry

I missed this notification... I am not a maintainer of this repo and can't merge prs, but can try to write some thoughts

P.S.
I think adding an {{ item }} would be a good idea:
- CONTRIBUTING.md
- PULL_REQUEST_TEMPLATE.md with clear steps creating prs and reviewing
- CODEOWNERS file
- Enable team pr [auto-assignment](https://docs.github.com/en/github/setting-up-and-managing-organizations-and-teams/managing-code-review-assignment-for-your-team)

**Comment by jgiannuzzi** [comment]
Note for reviewers: all tests are successful for this PR for all the combinations of OS (Linux, macOS, and Windows), .NET runtime (Core 2.1, Core 3.1, Framework 4.6.1, and .NET 5.0), and Consul (1.6, 1.7, 1.8, and 1.9) that we support.

See the results [here](https://github.com/jgiannuzzi/consuldotnet/actions/runs/685879394) under the preview version of our new CI.

**Comment by jgiannuzzi** [comment]
Hi @declard, could you please add filter tests that actually call the Consul agent?
You can grep for `Filter:` in the `api` subdirectory of the [Consul repo](https://github.com/hashicorp/consul) to find inspiration.

**Comment by declard** [comment]
> Hi @declard, could you please add filter tests that actually call the Consul agent?
> You can grep for `Filter:` in the `api` subdirectory of the [Consul repo](https://github.com/hashicorp/consul) to find inspiration.

Surely, will do

**Comment by declard** [comment]
Integration tests have been added

**Comment by jgiannuzzi** [comment]
The CI tests that are currently failing should be fixed by #74.

**Comment by jgiannuzzi** [comment]
Hi @declard, could you please update/rebase your PR to the latest `master`? This should solve the current CI failures.

**Comment by declard** [comment]
@jgiannuzzi the branch has been rebased

**Review by gordonhopes — APPROVED** [review]
Looks good to me and it seems to all run correctly locally.

**Review by maurosbarigia — COMMENTED** [review]
I think it would make sense to implement the Equals and the GetHashCode as the warnings suggest. It's jsut 2 lines of code:
```c#
public override bool Equals(object obj) => this == obj;
public override int GetHashCode() => HashCode.Combine(Prefix, Name);
```

**Comment by jgiannuzzi** [comment]
Thanks @gordonhopes and @maurosbarigia for reviewing!

Thanks @declard for adding this very cool functionality to Consul.NET, this is awesome!

Could you please just either add the 2 methods that @maurosbarigia suggested above to `ServiceMetaEntrySelector`, `ServiceSelector`, and `StringFieldSelector` — or alternatively explain why you did not implement them?
Once done, we can merge your PR 😄

**Comment by declard** [comment]
> I think it would make sense to implement the Equals and the GetHashCode as the warnings suggest. It's jsut 2 lines of code:
> 
> ```cs
> public override bool Equals(object obj) => this == obj;
> public override int GetHashCode() => HashCode.Combine(Prefix, Name);
> ```

I don't think it makes sense as the operators are overloaded exclusively to build an eDSL. It would be even worse if the class had `Equals` overridden as `==` and `!=` are completely unrelated to the equality relation on the type. It looks to me that the code analysis gives a false positive in this case.

**Comment by maurosbarigia** [comment]
I see what you mean. Thanks for explaining. I misinterpreted what you were using the == override for (and didn't notice it returns a Filter). I agree it doesn't need Equals and GetHashCode. @jgiannuzzi this is good to merge.

**Review by jgiannuzzi — APPROVED** [review]

**Pull request merged**
